codec/src/main/java/io/netty/handler/codec/ByteString.java

```diff
@@ -0,0 +1,770 @@
+        } else {
+            byte[] v = new byte[length];
+            int oldPos = value.position();
+            value.get(v, 0, length);
+            value.position(oldPos);
+            return v;
+        }
+    }
+
+    /**
+     * Create a copy of the underlying storage from {@link value} into a byte array.
+     * The copy will start at {@link ByteBuf#readerIndex()} and copy {@link ByteBuf#readableBytes()} bytes.
+     */
+    public static byte[] getBytes(ByteBuf buf) {
+        return getBytes(buf, checkNotNull(buf, "buf").readerIndex(), buf.readableBytes());
+    }
+
+    /**
+     * Create a copy of the underlying storage from {@link buf} into a byte array.
+     * The copy will start at {@code start} and copy {@code length} bytes.
+     */
+    public static byte[] getBytes(ByteBuf buf, int start, int length) {
+        return getBytes(buf, start, length, true);
+    }
+
+    /**
+     * Return an array of the underlying storage from {@link buf} into a byte array.
+     * The copy will start at {@code start} and copy {@code length} bytes.
+     * If {@code copy} is true a copy will be made of the memory.
+     * If {@code copy} is false the underlying storage will be shared, if possible.
+     */
+    public static byte[] getBytes(ByteBuf buf, int start, int length, boolean copy) {
+        if (start < 0 || length > checkNotNull(buf, "buf").capacity() - start) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= start(" + start + ") <= start + length(" + length
+                    + ") <= " + "buf.capacity(" + buf.capacity() + ')');
+        }
+
+        if (buf.hasArray()) {
+            if (copy || start != 0 || length != buf.capacity()) {
+                int baseOffset = buf.arrayOffset() + start;
+                return Arrays.copyOfRange(buf.array(), baseOffset, baseOffset + length);
+            } else {
+                return buf.array();
+            }
+        } else {
+            byte[] v = new byte[length];
+            buf.getBytes(start, v);
+            return v;
+        }
+    }
+
+    /**
+     * Create a copy of {@link value} into a byte array using the encoding type of {@code charset}.
+     */
+    public static byte[] getBytes(char[] value, Charset charset) {
+        return getBytes(value, charset, 0, checkNotNull(value, "value").length);
+    }
+
+    /**
+     * Create a copy of {@link value} into a byte array using the encoding type of {@code charset}.
+     * The copy will start at index {@code start} and copy {@code length} bytes.
+     */
+    public static byte[] getBytes(char[] value, Charset charset, int start, int length) {
+        if (start < 0 || length > checkNotNull(value, "value").length - start) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= start(" + start + ") <= start + length(" + length
+                    + ") <= " + "length(" + length + ')');
+        }
+        final ByteBuf buf = Unpooled.copiedBuffer(value, charset);
+        try {
```

Comment:
@normanmaurer is there any performance issue with wrapping the array with a ByteBuf, only to unwrap it after?